### ISSUE TITLE
QOS for late subscriber to recieve all messages

### ISSUE THREAD

**Issue by Titikshaaaa** [eclipse-cyclonedds/cyclonedds#1782]
I have set the writer qos as
dwQos = dds_qos_create();
dds_qset_reliability (dwQos, DDS_RELIABILITY_RELIABLE, DDS_INFINITY);
 dds_qset_durability(dwQos, DDS_DURABILITY_TRANSIENT_LOCAL);
dds_qset_history (dwQos, DDS_HISTORY_KEEP_ALL, 0);
dds_qset_resource_limits (dwQos, 100, DDS_LENGTH_UNLIMITED, DDS_LENGTH_UNLIMITED);
msg.userID = 1;
write message
printf ("=== [Publisher]  Writing : ");
 for (int i=0;i<50;i++){msg.message = "Hello world";
rc = dds_write (writer, &msg);
// DDS_ERR_CHECK (rc, DDS_CHECK_REPORT | DDS_CHECK_EXIT);
 printf ("Message (%d)\n", msg.userID);
 msg.userID++;
 dds_sleepfor (DDS_MSECS (1000));}
Now according to me if my subscriber has joined on Message (6) Then he should get all the previous message also.
But it is not working.. What am i missing

**Comment by eboasson** [comment]
The thing you've missed is that there is a separate setting history for the "historical data", the already-published data that a late-joining subscriber gets: the "durability service" QoS setting. I would expect you get everything if you add `dds_qset_durability_service(dwQos, 0, DDS_HISTORY_KEEP_ALL, 0, DDS_LENGTH_UNLIMITED, DDS_LENGTH_UNLIMITED, DDS_LENGTH_UNLIMITED)`.

Obviously setting a history that retains *everything* for all time can be tricky with resource usage. This is the reason you can set how much data is maintained for late-joining subscribers separately. For example, if you'd do `dds_qset_durability_service(dwQos, 0, DDS_HISTORY_KEEP_LAST, 10, ...)` instead, your subscriber gets the most recent 10 samples on joining, but it will still get everything published from then on because of `dds_qset_history (dwQos, DDS_HISTORY_KEEP_ALL, 0)`.

**Comment by Titikshaaaa** [comment]
Okay @eboasson I tried your solution but for me it isnt working after this also. Do i have to add any qos in subscriber side. My reader qos is
 qos = dds_create_qos ();
dds_qset_reliability (qos, DDS_RELIABILITY_RELIABLE, DDS_SECS (10)); 
 while (true)
  {
    /* Do the actual read.
     * The return value contains the number of read samples. */
    rc = dds_take(reader, samples, infos, MAX_SAMPLES, MAX_SAMPLES);
    if (rc < 0)
      DDS_FATAL("dds_read: %s\n", dds_strretcode(-rc));
    /* Check if we read some data and it is valid. */
    if ((rc > 0) && (infos[0].valid_data))
    {
      /* Print Message. */
      msg = (HelloWorldData_Msg*) samples[0];
      printf ("=== [Subscriber] Received : ");
      printf ("Message (%"PRId32", %s)\n", msg->userID, msg->message);
      fflush (stdout);
         dds_sleepfor (DDS_MSECS (1000));
    }
    else
    {
      dds_sleepfor (DDS_MSECS (20));
}} Am i missing something here??

**Comment by eboasson** [comment]
When I responded earlier I wondered whether I should note that you also have to configure the reader correctly, assumed you had and that your question was about why it only got 1 historical sample, because that question I have gotten a few times before ... Assumptions ...

The "transient-local" setting on the writer makes the writer keep historical data around for whoever needs it (with the "durability service" settings controlling how much is kept), the "transient-local" setting on the reader makes it request historical data.[^1]

If the writer is configured to retain the last 10 samples but the reader uses, e.g., a history setting of "keep last 1", then how many of those 10 samples you will actually see depends on various things, including whether there is a network in between and the timing of the reading in the application.

In short, you need at least
```
dds_qset_durability(qos, DDS_DURABILITY_TRANSIENT_LOCAL)
```
but you might also want to think about what history setting you'd like to use on the reader side for your experiment.

[^1]: The rules say that if the writer is "volatile" and the reader is "transient-local" the two don't communicate.

**Comment by Titikshaaaa** [comment]
Thanks a lot @eboasson .I got more clarity now and it is working for me.
